File: 2022/day04/day04.py
#!/usr/bin/env python3
import re
import sys

# comparing the start and end ranges is more efficient, but set operations require less thinking :)

def main():
    lines = sys.stdin.read().strip().split('\n')

    ans1 = 0
    for line in lines:
        a, b, c, d = map(int, re.split(r'[-,]', line))
        x = set(range(a, b + 1))
        y = set(range(c, d + 1))

        if x.issubset(y) or y.issubset(x):
            ans1 += 1

    print('part1:', ans1)

    ans2 = 0
    for line in lines:
        a, b, c, d = map(int, re.split(r'[-,]', line))
        x = set(range(a, b + 1))
        y = set(range(c, d + 1))

        if x & y:
            ans2 += 1

    print('part2:', ans2)

    assert ans1 == 595
    assert ans2 == 952


if __name__ == '__main__':
    main()

File: 2019/day04/day04.ts
import { readFileSync } from 'fs';

const regex = new RegExp('(\\d)\\1+');

function isValid1(s: string) {
  for (let i = 0; i < s.length - 1; i++)
    if (s[i] > s[i + 1])
      return false;
  return regex.test(s);
}

function isValid2(s: string) {
  if (!isValid1(s))
    return false;

  s = 'x' + s + 'x';
  for (let i = 0; i < s.length - 3; i++)
    if (s[i + 1] == s[i + 2] && s[i] != s[i + 1] && s[i + 2] != s[i + 3])
      return true;
  return false;
}

function main() {
  let [start, end] = readFileSync('input', 'utf-8').split('-').map(v => parseInt(v));
  let ans1 = 0, ans2 = 0;
  for (let n = start; n <= end; n++) {
    if (isValid1(String(n)))
      ans1++;
    if (isValid2(String(n)))
      ans2++;
  }

  console.log(`part1: ${ans1}`);
  console.log(`part2: ${ans2}`);
}

main();
